oak_functions/loader/src/lib.rs

```diff
@@ -164,76 +102,92 @@ async fn background_refresh_lookup_data(
 
 /// This crate is just a library so this function does not get executed directly by anything, it
 /// needs to be wrapped in the "actual" `main` from a bin crate.
-pub fn lib_main() -> anyhow::Result<()> {
-    let opt = Opt::parse();
-    let config_file_bytes = fs::read(&opt.config_path)
-        .with_context(|| format!("Couldn't read config file {}", &opt.config_path))?;
-    let config: Config =
-        toml::from_slice(&config_file_bytes).context("Couldn't parse config file")?;
-    // TODO(#1971): Make maximum log level configurable.
-    let logger = Logger::default();
-    logger.log_public(Level::Info, &format!("parsed config file:\n{:#?}", config));
+pub fn lib_main(
+    opt: Opt,
+    logger: Logger,
+    load_lookup_data_config: LoadLookupDataConfig,
+    worker_threads: Option<usize>,
+    policy: Option<Policy>,
+    extension_factories: Vec<Box<dyn ExtensionFactory<Logger>>>,
+    extension_configuration_info: ExtensionConfigurationInfo,
+) -> anyhow::Result<()> {
     tokio::runtime::Builder::new_multi_thread()
-        .worker_threads(config.worker_threads.unwrap_or(4))
+        .worker_threads(worker_threads.unwrap_or(4))
         .enable_all()
         .build()
         .unwrap()
-        .block_on(async_main(opt, config, logger))
+        .block_on(async_main(
+            opt,
+            logger,
+            load_lookup_data_config,
+            policy,
+            extension_factories,
+            extension_configuration_info,
+        ))
 }
 
-/// Main execution point for the Oak Functions Loader.
-async fn async_main(opt: Opt, config: Config, logger: Logger) -> anyhow::Result<()> {
-    let (notify_sender, notify_receiver) = tokio::sync::oneshot::channel::<()>();
-
-    let lookup_data_manager = load_lookup_data(&config, logger.clone()).await?;
-
-    #[allow(unused_mut)]
-    let mut extensions = Vec::new();
-
-    let workload_logging_factory =
-        WorkloadLoggingFactory::new_boxed_extension_factory(logger.clone())?;
-    extensions.push(workload_logging_factory);
-
-    let lookup_factory = LookupFactory::new_boxed_extension_factory(lookup_data_manager)?;
-    extensions.push(lookup_factory);
+/// Workaround to pass values for ConfigurationInfo from Config.
+pub struct ExtensionConfigurationInfo {
+    ml_inference: bool,
+    metrics: Option<oak_functions_abi::proto::PrivateMetricsConfig>,
+}
```

Comment:
+1.  These configurations should probably be added the extension factories rather than passed to the runtime. This is likely a significant refactor, so should probably be a separate PR.